src/swupd/hash.go

```diff
@@ -28,3 +38,88 @@ func (h hashval) String() string {
 func HashEquals(h1 hashval, h2 hashval) bool {
 	return h1 == h2
 }
+
+// Hashcalc returns the swupd hash for the given file
+func Hashcalc(filename string) (hashval, error) {
+	var info syscall.Stat_t
+	var err error
+	var data []byte
+	if err = syscall.Lstat(filename, &info); err != nil {
+		return 0, fmt.Errorf("Error stating file '%s' %v\n", filename, err)
+	}
+	// Get magic constants out of /usr/include/bits/stat.h
+	switch info.Mode & 0170000 {
+	case 0100000: // Regular file
+		data, err = ioutil.ReadFile(filename)
+		if err != nil {
+			return 0, fmt.Errorf("Read error for '%s' %v\n", filename, err)
+		}
+	case 0040000: // Directory
+		info.Size = 0
+		data = []byte("DIRECTORY") // fixed magic string
+	case 0120000:
```

Comment:
The os.Readlink clues me in but adding the comment for symlink to this line couldn't hurt.

Reply:
Will do. I don't like the magic numbers at all, they  make the program unportable, I did think, very briefly, of writing a C program to generate the values.